### PR TITLE
Support Canary Release

### PR DESCRIPTION
**Pull request by Bryce-huang**
https://github.com/OpenFunction/OpenFunction/issues/487

### CODE REVIEW

**Comment by Bryce-huang** [comment]
Don‘ t merge now.

**Comment by benjaminhuo** [comment]
The e2e test has know issues which is still failing, it's ok not to fix it now.
But it's better to fix https://github.com/OpenFunction/OpenFunction/actions/runs/6109785188/job/16581498056?pr=490 maybe :) @Bryce-huang

**Comment by benjaminhuo** [comment]
@Bryce-huang Is this ready for review?

**Comment by Bryce-huang** [comment]
> @Bryce-huang Is this ready for review?

Yes,I'm also doing tests.

**Comment by benjaminhuo** [comment]
> > @Bryce-huang Is this ready for review?
> 
> Yes,I'm also doing tests.

@wrongerror @wanjunlei @tpiperatgod @lizzzcai would you help to review this?

File: apis/core/v1beta1/function_types.go [hunk 1]
```diff
@@ -20,7 +20,7 @@ import (
 	componentsv1alpha1 "github.com/dapr/dapr/pkg/apis/components/v1alpha1"
 	v1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	k8sgatewayapiv1alpha2 "sigs.k8s.io/gateway-api/apis/v1alpha2"
+	k8sgatewayapiv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
```
Comment (wrongerror) [inline apis/core/v1beta1/function_types.go:23]: Old versions of the API ( v1beta1 ) don't need updating anymore?

Comment (benjaminhuo) [reply apis/core/v1beta1/function_types.go:23]: To use gateway API v1beta1, what's the minimum version of contour we should use?

Comment (Bryce-huang) [reply apis/core/v1beta1/function_types.go:23]: Covers using version v1beta1 of the Gateway API, with Contour v1.22.0 or higher.

https://projectcontour.io/guides/gateway-api/

**Comment by benjaminhuo** [comment]
I would suggest the following changes:
@Bryce-huang @wrongerror @wanjunlei @tpiperatgod @lizzzcai what do you think?

```yaml
apiVersion: core.openfunction.io/v1beta2
kind: Function
metadata:
  name: test-server
spec:
  rolloutStrategy:
    canary:
      steps:
      - weight: 20
        pause:
          duration: 60
      - weight: 80
         pause:
          duration: 120
  image: brycehuang/web-service-image:v1
  serving:
    template:
      containers:
      - imagePullPolicy: IfNotPresent
        name: function
  version: latest
  workloadRuntime: OCIContainer
status:
  # status.serving and status.revision should always be the stable serving and revision
  serving:
  revision: 
  rollout:
    canary:
      status: # CanaryStatus
      canaryRevision:
	serving:
	revision
```

**Comment by Bryce-huang** [comment]
> I would suggest the following changes: @Bryce-huang @wrongerror @wanjunlei @tpiperatgod @lizzzcai what do you think?
> 
> ```yaml
> apiVersion: core.openfunction.io/v1beta2
> kind: Function
> metadata:
>   name: test-server
> spec:
>   rolloutStrategy:
>     canary:
>       steps:
>       - weight: 20
>         pause:
>           duration: 60
>       - weight: 80
>         pause:
>           duration: 120
>   image: brycehuang/web-service-image:v1
>   serving:
>     template:
>       containers:
>       - imagePullPolicy: IfNotPresent
>         name: function
>   version: latest
>   workloadRuntime: OCIContainer
> status:
>   # status.serving and status.revision should always be the stable serving and revision
>   # status.canary.serving and status.canary.revision should be the canary ones
>   serving:
>   revision: 
>   rollout:
>     canary:
>       status: # CanaryStatus
>       serving:
>       revision:
> ```



> I would suggest the following changes: @Bryce-huang @wrongerror @wanjunlei @tpiperatgod @lizzzcai what do you think?
> 
> ```yaml
> apiVersion: core.openfunction.io/v1beta2
> kind: Function
> metadata:
>   name: test-server
> spec:
>   rolloutStrategy:
>     canary:
>       steps:
>       - weight: 20
>         pause:
>           duration: 60
>       - weight: 80
>         pause:
>           duration: 120
>   image: brycehuang/web-service-image:v1
>   serving:
>     template:
>       containers:
>       - imagePullPolicy: IfNotPresent
>         name: function
>   version: latest
>   workloadRuntime: OCIContainer
> status:
>   # status.serving and status.revision should always be the stable serving and revision
>   # status.canary.serving and status.canary.revision should be the canary ones
>   serving:
>   revision: 
>   rollout:
>     canary:
>       status: # CanaryStatus
>       serving:
>       revision:
> ```

LGTM

File: controllers/core/function_controller.go [hunk 2]
```diff
@@ -127,19 +132,125 @@ func (r *FunctionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+		if err := r.cleanServing(fn); err != nil {
+			log.Error(err, "Failed to clean Serving")
+			return nil, err
+		}
+	}
+
+	return recheckTime, nil
+}
+func hasCanaryReleasePlan(fn *openfunction.Function) bool {
+	if len(fn.Spec.RolloutStrategy.Canary.Steps) == 0 {
```
Comment (benjaminhuo) [inline controllers/core/function_controller.go:249]: should check if RolloutStrategy is defined first?

**Review by wanjunlei — CHANGES_REQUESTED** [review]

**Pull request closed without merging**